BCH integration

DarrenWestwood opened the pull request:
- Removed Flyp integration
- Added Blockonomics BCH integration

shivaenigma commented:
https://prnt.sc/r8u831

blockonomics commented:
Address reuse is not working in BTC. Will test further after this is fixed

blockonomics commented:
Let us use simple encryption to generate uuid from encrypt(order_id, key) instead of saving to DB
https://stackoverflow.com/a/2490376/5043010

blockonomics commented:
Change this for BCH payment https://prnt.sc/ravm7x

blockonomics commented:
Let us use the below design for switcher screen. Also lets remove heading *Bitcoin/Bitcoin Cash Payment* and icon from the actual payment screen

https://www.figma.com/proto/GSOPajdDI8mWMORdCZ7NXopp/Blockonomics?node-id=966%3A638&viewport=234%2C284%2C0.568193256855011&scaling=min-zoom

DarrenWestwood commented:
Updated with the new ui changes and encryption of order id instead of using uuid

shivaenigma commented on modules/gateways/callback/blockonomics.php:
```diff
@@ -47,20 +47,26 @@
 $confirmations = $blockonomics->getConfirmations();
 
+$blockonomics_currency_code = $order['blockonomics_currency'];
+$active_currencies = $blockonomics->getActiveCurrencies()['currencies'];
```
Instead of *getActiveCurrencies* this can use *getSupportedCurrencies* . *getActiveCurrencies* has overhead of extra DB call that is not required here.

DarrenWestwood replied:
Fixed with dbcfd12

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -79,19 +79,55 @@ private function generateCallbackSecret() {
-			->where('gateway', 'blockonomics')
-			->where('setting', 'Altcoins')
-			->value('value');
+	public function getSupportedCurrencies() {
```
This can return a dict() so that a caller that can iterating over array to match code

DarrenWestwood replied:
Fixed with 5138aa5

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -79,19 +79,55 @@ private function generateCallbackSecret() {
+  				)
+  			)
+  		);
+		return json_encode($all_currencies);
```
Why have overhead of *json_encode* when all calling functions are doing *json_decode* ?

DarrenWestwood replied:
Removed in 5138aa5

shivaenigma commented:
![Screenshot from 2020-03-23 21-29-30](https://user-images.githubusercontent.com/737092/77336404-77521980-6d4d-11ea-9162-e6e5cefe9b37.png)

For BCH ... Put this: 
Note to webmaster: Please follow instructions here (freshdesk article link)  to configure BCH payments

shivaenigma commented:
https://github.com/blockonomics/whmcs-bitcoin-plugin/blob/b56806157a323b34000c3843badc07e306347786/templates/blockonomics/payment.tpl#L16

This is still linking to blockonomics.co even in case of BCH txid. Its just easier actually to remove the txid from this line, since the txid can anyway be seen on the invoice

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+
+        return bin2hex($iv . $hash . $cipherText);
+    }
+
```
The order_id is becoming very large and such security is not required here.
- Let us use aes-128-cbc
- iv,  use same length as callback secret truncate/padded to 16 bytes. No need to pass it
- No need to to hmac and send it. If even attacker puts in random string it is very less likely it will decrypt to an actual order_id

This is make the encrypt order_id to be 16bytes that is enough

DarrenWestwood replied:
Changed in 7bcabc6

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -204,9 +245,11 @@ public function getNewBitcoinAddress($reset=false) {
 		curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
 		curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
+		curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 20); 
+		curl_setopt($ch, CURLOPT_TIMEOUT, 20); //timeout in seconds
```
Twenty seconds timeout is very large! Anyway we haven't seen any address generation timeout issues on WHMCS. Let us remove this

DarrenWestwood replied:
Removed in 97b9f5c

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -234,14 +278,22 @@ public function getMargin() {
 	 * Convert fiat amount to BTC
 	 */
-	public function getBitcoinAmount($fiat_amount, $currency) {
+	public function getNewAmount($fiat_amount , $blockonomics_currency = 'btc') {
```
Please choose a better function name. It is not very clear what this function is doing

DarrenWestwood replied:
Changed in d3329f4

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -254,11 +306,12 @@ public function getBitcoinAmount($fiat_amount, $currency) {
 		}
 
-		return intval(1.0e8 * $fiat_amount/$price);
+		return;
```
returning empty is a bad design. Best to let the exception propagate to parent or return an invalid value that the parent will check

DarrenWestwood replied:
Moved in a4a5005

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -234,14 +278,22 @@ public function getMargin() {
 	 */
-	public function getBitcoinAmount($fiat_amount, $currency) {
+	public function getNewAmount($fiat_amount , $blockonomics_currency = 'btc') {
+		$currency = getCurrency(getClientsDetails()['user_id'])['code'];
```
why do we need an additional call ? I think the caller function will have the order currency. Why not pass that too ?

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -279,23 +332,26 @@ public function createOrderTableIfNotExist() {
+		}else{
+			if(!Capsule::schema()->hasColumn('blockonomics_bitcoin_orders', 'blockonomics_currency')){
+				 Capsule::schema()->table('blockonomics_bitcoin_orders', function($table){
+					$table->string('blockonomics_currency');
```
We may need to set default value of this to 'btc' to take care of existing invoices/ invoices with ongoing payments. 
Also lets remove flyp_id column if it exists

DarrenWestwood replied:
Removed flyp_id in 04ed4ca

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
 				->first();
 		} catch (\Exception $e) {
-				echo "Unable to select order from blockonomics_bitcoin_orders: {$e->getMessage()}";
+				echo "Unable to get order id from blockonomics_bitcoin_orders: {$e->getMessage()}";
```
What is the use of the select when decrypt itself is giving order_id ?

DarrenWestwood replied:
Returns the full order details from the order_id. Updated the Exception message in 9dcba72

shivaenigma replied:
It is returning `return $first_order->id_order` ..  ?

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+	/*
+	 * Get existing or create new order
+	 */	
+	public function getOrderByHash($order_uuid, $blockonomics_currency) {
```
- Please rename method name, getOrder implies it is only fetching details. However, this is doing lots of stuff like generating address / setting price
- Rename *order_uuid* variable to order hash as we don't have uuid anymore

DarrenWestwood replied:
Updated in 7cf8d9f

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+		try {
+			return Capsule::table('blockonomics_bitcoin_orders')
+				->where('id_order', $order_id)
+				->orderBy('timestamp', 'desc')->get();
```
Order by timestamp is unnecessary as it creates additional overhead on query

DarrenWestwood replied:
Removed in 7e4b565

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+				return $order;
+			}
+		}
+		return;
```
Empty return value ?

DarrenWestwood replied:
Removed in 4b674d6

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+	/*
+	 * Check for unused address for the currency linked to the order
+	 */		
+	public function isOrderWaiting($orders, $blockonomics_currency) {
```
Again.. rename function since does also does updates and not only doing a get. Something like *getAndUpdateWaitingOrder*

DarrenWestwood replied:
Updated in 18ea106

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+	 */	
+	public function getOrderByHash($order_uuid, $blockonomics_currency) {
+		// Fetch all orders by uuid
+		$all_orders_by_uuid = $this->getAllOrdersByHash($order_uuid);
```
Make the getAllOrdersbyHash return a dict whose key is blockonomics_currency. This way all the following functions can avoid uncessary lops

DarrenWestwood replied:
There can be multiple addresses with the same blockonomics_currency, due to multiple payments for a single invoice.

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+			return $pending_payment;
+		}
+		// Check for existing address
+		$address_waiting = $this->isOrderWaiting($all_orders_by_uuid, $blockonomics_currency);
```
Make isOrderWaiting take single order as input that is of the type $blockonomics_currency

DarrenWestwood replied:
Depends upon
> Make the getAllOrdersbyHash return a dict

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+				->orderBy('timestamp', 'desc')->get();
+		return $all_orders_by_id;
+	}
+
```
Unused function? Please remove

DarrenWestwood replied:
Removed in dc3500c

shivaenigma commented:
- Remove all calls to getCurrency() 
- We need to store the fiat currency in DB as a column.. like we are doing in wordpress
- Set the fiat currency value also when order is created (it is insert the fiat order value)

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+			return $address_waiting;
+		}
+		// Check for new order
+		$new_order = $this->isOrderNew($all_orders_by_uuid, $blockonomics_currency);
```
Pass isOrderNew a single order as input that already has the correct $blockonomics_currency

DarrenWestwood replied:
Depends upon
> Make the getAllOrdersbyHash return a dict

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -327,39 +384,241 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
-	 * Try to get order row from db by uuid
+	 * Check for unused address or create new
+	 */	
+	public function isOrderNew($orders, $blockonomics_currency) {
```
Again choose a bettter function name. _isOrderNew_ is only supposed to do a check request and return a flag.

DarrenWestwood replied:
Updated in 18ea106

DarrenWestwood commented:
> Note to webmaster: Please follow instructions here (freshdesk article link) to configure BCH payments

Added with a739795

> This is still linking to blockonomics.co even in case of BCH txid. Its just easier actually to remove the txid from this line, since the txid can anyway be seen on the invoice

Removed with 05725b1

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -79,19 +79,51 @@ private function generateCallbackSecret() {
+					->where('gateway', 'blockonomics')
+					->where('setting', 'ApiKey')
+					->value('value');
+			}else{
```
Let us remove this. Keep btc always part of active currency

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -279,30 +325,38 @@ public function createOrderTableIfNotExist() {
 							$table->integer('bits_payed');
-							$table->text('flyp_id');
+							$table->string('blockonomics_currency');
 						}
```
For existing orders (before upgrade), we should set this a default value as BTC.

DarrenWestwood replied:
Added default for existing orders in 5fb998d

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -79,19 +79,51 @@ private function generateCallbackSecret() {
+				$api = Capsule::table('tblpaymentgateways')
+					->where('gateway', 'blockonomics')
+					->where('setting', 'ApiKey')
+					->value('value');
```
Remove this. Not necessary. Keep BTC always as active currency

DarrenWestwood replied:
Changed in af03268

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -279,30 +325,38 @@ public function createOrderTableIfNotExist() {
 	 * If order exists, return with false
 	 */
-	public function insertOrderToDb($id_order, $address, $value, $bits) {
+	public function insertOrderToDb($id_order, $address, $value, $bits, $blockonomics_currency = 'btc') {
```
Unused function. Please remove

DarrenWestwood replied:
Removed in df70277

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -318,63 +372,243 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+			return $address_waiting;
+		}
+		// Process a new order for the hash and blockonomics currency
+		$new_order = $this->processNewOrder($all_orders_by_hash, $blockonomics_currency);
```
Comments on processNewOrder. The whole process of creating order with empty address is unnecessary  as it creates empty rows. Let us do the following:
-  Rename *processNewOrder* to *createNewCryptoOrder* and pass it order_id, order_value and blockonomics_currency. No need to loop just take these values from $all_orders_by_hash[0]
-  *updateOrderAddress* will first search order having empty  addr. If found it will use that, else it will create new row with iinput order_value, order_id and address and blockonomics_currency

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -318,63 +372,243 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+		$order = $this->processOrderHash($order_uuid, $blockonomics_currency);
+		return $order;
+	}
+
```
I am not sure why these lines are needed. This creates a potentially recursive loop since processOrderHash calls this function again

DarrenWestwood replied:
These lines are required when the user has generated an address already, but did not pay, then selects another currency. There is no skeleton order waiting. Updated to prevent possible loop issues in 96d9dd4

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -318,63 +372,243 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+	 * Add a new order to the db
 	 */
-	public function getOrderByAddress($bitcoinAddress) {
+	public function newOrder($amount, $id_order) {
```
Rename to *createSkeletonOrder*

DarrenWestwood replied:
Renamed in 133d377

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -318,63 +372,243 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+	/*
+	 * Check for unused address or create new
+	 */	
+	public function processNewOrder($orders, $blockonomics_currency) {
```
Rename to *createNewCryptoOrder*

DarrenWestwood replied:
Rensamed in bbcb093

shivaenigma commented:
The whole process of converting the POST payment to GET isn't a very trivial ones. Coupled this with creating a new orderhash, we should have edited the design documen before making any code changes. *This is why this is taking lot of time and rework* . We should have a good design first covering all the cases.

About inserting order_id, value and timestamp with empty address, and then searching it again to find the order_value creates problems. Let us see this case:

1. Customer creates order and attempts pay (but doesn't select cryptocurrency)
1. Now he gets credits via PayPal
1. Now if he does checkout there are two rows in DB with empty *addr* and it creates ambiguity on which to generate address for. If he selected say BTC in step 1, it creates another problem as now the invoice will try to pay the value of the previous order (before the credits) due to our address reuse logic

DarrenWestwood commented:
> About inserting order_id, value and timestamp with empty address, and then searching it again to find the order_value creates problems.

We can sort the orders by timestamp or adjust to the correct logic to reuse unused addresses for the same order, even if the amount has changed.

shivaenigma commented:
> > About inserting order_id, value and timestamp with empty address, and then searching it again to find the order_value creates problems.
> 
> We can sort the orders by timestamp or adjust to the correct logic to reuse unused addresses for the same order, even if the amount has changed.

This creates more complexity. Also it is better to have deterministic rather than guessing the order from latest timestamp. Here is a possible solution:

- order hash will be encrypt(order_id:order_value)
- This way order hash itself has the order_value and there is no need to insert orders with empty address and unnecessary rows. 
- Remove *createSkeletonOrder* and *getSkeletonOrderByHash*
- Simplify *processOrderHash* and *createNewCryptoOrder* methods (removed unecessary code and loops)

shivaenigma commented:
@DarrenWestwood  Please ping me on slack before taking up any further changes on this

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -2,599 +2,753 @@
+                    } catch (Exception $e) {
+                        exit("Unable to set default value for existing blockonomics_bitcoin_orders: {$e->getMessage()}");
+                    }
+                });
```
This is necessary. There is a simple modifier to set default value https://laravel.com/docs/5.2/migrations#creating-columns

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -279,148 +286,258 @@ public function createOrderTableIfNotExist() {
 							$table->integer('bits');
 							$table->integer('bits_payed');
-							$table->text('flyp_id');
+							$table->string('blockonomics_currency');
```
Create new table `blockonomics_orders` . Let us not worry about migration. 
- We will not have `id` column
- `addr` will be primary_key
- add index to `id_order` column (https://laravel.com/docs/5.2/migrations#creating-indexes )

DarrenWestwood replied:
Updated in ffcbf89 and 2bc64ea

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -279,148 +286,258 @@ public function createOrderTableIfNotExist() {
+	/*
+	 * Check for unused address for the currency linked to the order
+	 */		
+    public function getAndUpdateWaitingOrder($orders, $currency, $blockonomics_currency)
```
We should be passing *value* also as parameter and use this the waiting order value. 
Use case;
1. customer has paid a previous watiing order by paypal
2. new order to make partial payment now reuseses existing address (but different value)

DarrenWestwood replied:
Added in 84f4203

shivaenigma replied:
The parameter names *orders* and *order_info* together are confusing. Use better names that reflect usage. Also update function description comment to better describe working

DarrenWestwood replied:
Updated in c0283b4

shivaenigma commented on modules/gateways/Blockonomics/Blockonomics.php:
```diff
@@ -279,148 +286,258 @@ public function createOrderTableIfNotExist() {
+                //Check if existing order is expired
+                $current_time = time();
+                $total_time = $this->getTimePeriod() * 60;
+                $clock = $order->timestamp + $total_time - $current_time;
+                if ($clock < 0) {
+                    $order->bits = $this->convertFiatToBlockonomicsCurrency($order->value, $currency, $order->blockonomics_currency);
+                    $order->timestamp = $current_time;
+                }
```
Always refresh with latest timestamp. Don't worry about expiry

DarrenWestwood replied:
Updated in b5f5be1

shivaenigma commented on payment.php:
```diff
@@ -24,158 +24,58 @@
+		exit(json_encode($existing_order));
 	}
+}else if($finish_order){
+	$existing_order = $blockonomics->getOrderByAddress($finish_order);
```
finish_order should directly take order_id as parameter. Then we can delete getOrderByAddress function also

DarrenWestwood replied:
Updated in a74c4e5

shivaenigma replied:
Please remove *getOrderByAddress* as it is now unused

DarrenWestwood replied:
Function is still used in the callback file:
https://github.com/blockonomics/whmcs-bitcoin-plugin/blob/master/modules/gateways/callback/blockonomics.php#L44

shivaenigma commented:
Done Testing. All ok. Found a minor issue with unconfirmed tx URL. Please fix
![Screenshot from 2020-04-02 11-23-08](https://user-images.githubusercontent.com/737092/78215123-6c795080-74d4-11ea-82a2-c6be704e557d.png)

shivaenigma commented:
TODO before release: Add link to help article here https://prnt.sc/rrbo17

shivaenigma commented:
Remove this text: https://prnt.sc/rrbslq

DarrenWestwood commented:
> Found a minor issue with unconfirmed tx URL.
Fixed in 80d6956 and 6633044

> TODO before release: Add link to help article here https://prnt.sc/rrbo17
Updated in c0f7d49

>Remove this text: https://prnt.sc/rrbslq
Removed in d385d63

jusasiiv commented on js/app.js:
```diff
@@ -117,382 +84,155 @@ app.controller('CheckoutController', function($scope, $interval, Order, $httpPar
+    }
+    
+    //Copy bitcoin address to clipboard
+    $scope.blockonomics_address_click = function() {
```
This and `blockonomics_amount_click()` have the same code except element id. Maybe pass `divid` as argument for the function

DarrenWestwood replied:
Updated in b6ffe22

shivaenigma commented:
Are cryptofont files used now in BCH integration ? Please remove if not used

DarrenWestwood commented:
> I think this function `$scope.alt_track_url` is not used anywhere anymore? (Line 64 of app.js if link does not work) https://github.com/blockonomics/whmcs-bitcoin-plugin/pull/70/files#diff-41d794d24ff042b1f9ac211fc3f9f951R64
> 
> Also this function `getParameterByNameBlocko()` (line 19) https://github.com/blockonomics/whmcs-bitcoin-plugin/pull/70/files#diff-41d794d24ff042b1f9ac211fc3f9f951R19

Removed with fccb919

> Are cryptofont files used now in BCH integration ? Please remove if not used

Removed with 7d23e5e
